engine/src/main/java/io/zeebe/engine/nwe/behavior/BpmnWorkflowResultSenderBehavior.java

```diff
@@ -0,0 +1,102 @@
+import io.zeebe.engine.state.instance.ElementInstanceState;
+import io.zeebe.engine.state.instance.VariablesState;
+import io.zeebe.protocol.impl.record.value.workflowinstance.WorkflowInstanceResultRecord;
+import io.zeebe.protocol.record.ValueType;
+import io.zeebe.protocol.record.intent.WorkflowInstanceResultIntent;
+import io.zeebe.protocol.record.value.BpmnElementType;
+import java.util.HashSet;
+import java.util.Set;
+import org.agrona.DirectBuffer;
+
+public final class BpmnWorkflowResultSenderBehavior {
+
+  private final WorkflowInstanceResultRecord resultRecord = new WorkflowInstanceResultRecord();
+
+  private final ElementInstanceState elementInstanceState;
+  private final VariablesState variablesState;
+  private final TypedResponseWriter responseWriter;
+
+  public BpmnWorkflowResultSenderBehavior(
+      final ZeebeState zeebeState, final TypedResponseWriter responseWriter) {
+    elementInstanceState = zeebeState.getWorkflowState().getElementInstanceState();
+    variablesState = elementInstanceState.getVariablesState();
+    this.responseWriter = responseWriter;
+  }
+
+  public void sendResult(final BpmnElementContext context) {
+
+    if (context.getBpmnElementType() != BpmnElementType.PROCESS) {
+      throw new BpmnProcessingException(
+          context,
+          "Expected to send the result of the workflow instance but was not called from the process element");
+    }
+
+    final AwaitWorkflowInstanceResultMetadata requestMetadata =
+        elementInstanceState.getAwaitResultRequestMetadata(context.getWorkflowInstanceKey());
+
+    if (requestMetadata != null) {
+      sendResult(context, requestMetadata);
+    }
+  }
+
+  private void sendResult(
+      final BpmnElementContext context, final AwaitWorkflowInstanceResultMetadata requestMetadata) {
+
+    final DirectBuffer variablesAsDocument = collectVariables(context, requestMetadata);
+
+    resultRecord
+        .setWorkflowInstanceKey(context.getWorkflowInstanceKey())
+        .setWorkflowKey(context.getWorkflowKey())
+        .setBpmnProcessId(context.getBpmnProcessId())
+        .setVersion(context.getWorkflowVersion())
+        .setVariables(variablesAsDocument);
+
+    responseWriter.writeResponse(
+        context.getWorkflowInstanceKey(),
+        WorkflowInstanceResultIntent.COMPLETED,
+        resultRecord,
+        ValueType.WORKFLOW_INSTANCE_RESULT,
+        requestMetadata.getRequestId(),
+        requestMetadata.getRequestStreamId());
+
+    responseWriter.flush();
+  }
+
+  private DirectBuffer collectVariables(
+      final BpmnElementContext context, final AwaitWorkflowInstanceResultMetadata requestMetadata) {
+
+    final Set<DirectBuffer> variablesToCollect = new HashSet<>();
+    requestMetadata
+        .fetchVariables()
+        .forEach(
+            variable -> {
+              final var variableName = cloneBuffer(variable.getValue());
+              variablesToCollect.add(variableName);
```

Comment:
I now notice theres mapping happening before the add, which can be moved before the foreach. It would even be possible to collect(toSet).

Reply:
Unfortunately not so easy :sweat_smile: 

`fetchVariables()` returns `ArrayProperty<StringValue>` and `ArrayProperty` implements only `Iterable`. I would need to use `StreamSupport` to create a stream of it. This seems a bit too much for it.

Reply:
Ah oops, I forgot about that.